segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,263 @@
+
+    private CompletableFuture<Void> loopFuture;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param chunkedSegmentStorage Instance of {@link ChunkedSegmentStorage}.
+     * @param config Configuration to use.
+     */
+    public GarbageCollector(ChunkedSegmentStorage chunkedSegmentStorage, ChunkedSegmentStorageConfig config) {
+        super("io.pravega.segmentstore.storage.chunklayer.GarbageCollector", (ScheduledExecutorService) chunkedSegmentStorage.getExecutor());
+        this.chunkedSegmentStorage = Preconditions.checkNotNull(chunkedSegmentStorage, "chunkedSegmentStorage");
+        this.config = Preconditions.checkNotNull(config, "config");
+    }
+
+    /**
+     * Initializes this instance.
+     */
+    public void initialize() {
+        Services.startAsync(this, this.executor);
+    }
+
+    /**
+     * Gets a value indicating how much to wait for the service to shut down, before failing it.
+     *
+     * @return The Duration.
+     */
+    @Override
+    protected Duration getShutdownTimeout() {
+        return null;
+    }
+
+    /**
+     * Main execution of the Service. When this Future completes, the service auto-shuts down.
+     *
+     * @return A CompletableFuture that, when completed, indicates the service is terminated. If the Future completed
+     * exceptionally, the Service will shut down with failure, otherwise it will terminate normally.
+     */
+    @Override
+    protected CompletableFuture<Void> doRun() {
+        loopFuture = Futures.loop(
+                this::canRun,
+                () -> CompletableFuture.completedFuture(null)
+                        .thenComposeAsync( v -> deleteGarbage(true, config.getGarbageCollectionConcurrency()), executor)
+                        .handleAsync((v, ex) -> {
+                            if (null != ex) {
+                                log.error("{} Error during run.", chunkedSegmentStorage.getLogPrefix(), ex);
+                            }
+                            return null;
+                        }, executor),
+                executor);
+        return loopFuture;
+    }
+
+    private boolean canRun() {
+        return isRunning() && getStopException() == null && !closed.get();
+    }
+
+    /**
+     * Sets whether background cleanup is suspended or not.
+     *
+     * @param value Boolean indicating whether to suspend background processing or not.
+     */
+    void setSuspended(boolean value) {
+        suspended.set(value);
+    }
+
+    /**
+     * collect the garbage chunks.
+     *
+     * @param chunksToDelete List of chunks to delete.
+     */
+    void addToGarbage(Collection<String> chunksToDelete) {
+        val currentTime = System.currentTimeMillis();
+
+        if (queueSize.get() < config.getGarbageCollectionMaxQueueSize()) {
+            chunksToDelete.forEach(chunkToDelete -> garbageChunks.add(new GarbageChunkInfo(chunkToDelete, currentTime)));
+            queueSize.incrementAndGet();
```

Comment:
We discussed that this is a bug, as you are not incrementing the size of the queue inside the loop, right?

Reply:
Please, add a test for the enforcement of queue size limits of Garbage Collector, as the fact that no test is failing due to this makes me feel that it needs better coverage.

Reply:
yes it was a silly mistake and tests didn't catch it :(
Added new tests.